File: server/internal/api/v1/relation.go
package api

import (
	"github.com/gin-gonic/gin"
	"interastral-peace.com/alnitak/internal/domain/dto"
	"interastral-peace.com/alnitak/internal/resp"
	"interastral-peace.com/alnitak/internal/service"
	"interastral-peace.com/alnitak/utils"
)

// 关注
func Follow(ctx *gin.Context) {
	var idReq dto.IdReq
	if err := ctx.Bind(&idReq); err != nil {
		resp.FailWithMessage(ctx, "请求参数有误")
		return
	}

	if err := service.Follow(ctx, idReq.ID); err != nil {
		resp.FailWithMessage(ctx, err.Error())
		return
	}

	// 返回
	resp.Ok(ctx)
}

// 取消关注
func Unfollow(ctx *gin.Context) {
	var idReq dto.IdReq
	if err := ctx.Bind(&idReq); err != nil {
		resp.FailWithMessage(ctx, "请求参数有误")
		return
	}

	if err := service.Unfollow(ctx, idReq.ID); err != nil {
		resp.FailWithMessage(ctx, err.Error())
		return
	}

	// 返回
	resp.Ok(ctx)
}

// 获取用户关系
func GetUserRelation(ctx *gin.Context) {
	targetUid := utils.StringToUint(ctx.Query("userId"))

	relation := service.GetUserRelation(ctx, targetUid)

	// 返回
	resp.OkWithData(ctx, gin.H{"relation": relation})
}

// 获取关注列表
func GetFollowings(ctx *gin.Context) {
	userId := utils.StringToUint(ctx.Query("userId"))
	page := utils.StringToInt(ctx.Query("page"))
	pageSize := utils.StringToInt(ctx.Query("pageSize"))

	if pageSize > 50 {
		resp.FailWithMessage(ctx, "请求数量过多")
		return
	}

	users, err := service.GetFollowings(ctx, userId, page, pageSize)
	if err != nil {
		resp.FailWithMessage(ctx, err.Error())
		return
	}

	// 返回给前端
	resp.OkWithData(ctx, gin.H{"users": users})
}

// 获取粉丝列表
func GetFollowers(ctx *gin.Context) {
	userId := utils.StringToUint(ctx.Query("userId"))
	page := utils.StringToInt(ctx.Query("page"))
	pageSize := utils.StringToInt(ctx.Query("pageSize"))

	if pageSize > 50 {
		resp.FailWithMessage(ctx, "请求数量过多")
		return
	}

	users, err := service.GetFollowers(ctx, userId, page, pageSize)
	if err != nil {
		resp.FailWithMessage(ctx, err.Error())
		return
	}

	// 返回给前端
	resp.OkWithData(ctx, gin.H{"users": users})
}

// 获取关注和粉丝数
func GetFollowCount(ctx *gin.Context) {
	userId := utils.StringToUint(ctx.Query("userId"))

	following, follower := service.GetFollowCount(ctx, userId)

	// 返回给前端
	resp.OkWithData(ctx, gin.H{"following": following, "follower": follower})
}

File: web/mobile-client/src/api/relation.ts
import request from '@/utils/request';
import { baseURL } from '@/utils/request';
import { useAsyncData } from 'nuxt/app';

// 获取用户关系
export const getUserRelationAPI = (userId: number) => {
  return request.get(`v1/relation/getUserRelation?userId=${userId}`);
}

// 关注
export const followAPI = (userId: number) => {
  return request.post("v1/relation/follow", { id: userId });
}

// 取消关注
export const unfollowAPI = (userId: number) => {
  return request.post("v1/relation/unfollow", { id: userId });
}

// 获取关注数据
export const getFollowDataAPI = (uid: number | string) => {
  return request.get(`v1/relation/getFollowCount?userId=${uid}`)
}

// 获取关注数据
export const asyncetFollowDataAPI = async (userId: number | string) => {
  return await useAsyncData(() => $fetch(`${baseURL}/api/v1/relation/getFollowCount?userId=${userId}`));
}

//获取关注列表
export const getFollowingAPI = (userId: number, page: number, pageSize: number) => {
  return request.get(`v1/relation/getFollowings?userId=${userId}&page=${page}&pageSize=${pageSize}`);
}

//获取粉丝列表
export const getFollowersAPI = (userId: number, page: number, pageSize: number) => {
  return request.get(`v1/relation/getFollowers?userId=${userId}&page=${page}&pageSize=${pageSize}`);
}